doc/design/migration_to_python.md

```diff
@@ -0,0 +1,479 @@
+  - The configure mechanism should be described in detail in the detailed design doc (the current architecture use the environment variable `SQLFlow_submitter` to config the platform)
+  - The locating mechanism should be described in detail in the detailed design doc.
+-  Call one of `train`, `predict`, `explain` ,`run`, `solve`, `evaluate` of the python module of the platform according to `statement.type`
+
+There should be at least three modules in the `platform` package:
+
+- `pai.py`
+- `alisa.py`
+- `default.py`
+
+The functions `train`, `predict`, `explain` ,`run`, `solve`, `evaluate` provided by each module in `platform` should forward to `sqlflow_sumitter/xgboost`, `sqlflow_submitter/tensorflow`, etc. according to `statement.type` and `statement.estimator`. **Diagnostics are required**.
+
+The functions `train`, `predict`, `explain` ,`run`, `solve`, `evaluate `  provided by each module in `platform` should have the same signature. The signature should be defined in the detailed design doc. **Diagnostics are required**.
+
+###### Golang
+
+To execute a SQLFlow program, the golang package `pkg/executor` should eventually spawn a python process that calls `execute.py:main("execute")`, and passes the filled `Program` to stdin of the process via the standard IPC pipe.
+
+To run a step of a executing SQLFlow program, the golang package `pkg/step` should spawn a python process that calls `execute.py:main("step")`, and passes the filled `Program` (with only one `Statement`) to stdin of the process via the standard IPC pipe.
+
+To get the status of an executing SQLFlow program, the golang package `pkg/server` should eventually spawn a python process that calls `execute.py:main("fetch")` .
+
+###### MVP
+
+The python function `fetch` is **not** required in the MVP.
+
+#### Feature Derivation And Columns
+
+###### Python
+
+The function calls in the COLUMN clause should directly map to calls to python functions in the `columns` package, for example, the `COLUMN NUMERIC(x)` statement would directly map to the python function `numeric()`. We should have at least a `tf_columns.py` module under this package to implement currently supported columns such as `numeric` or `indicator`.
+
+`features.py` should have at least two function: 
+
+```python
+# Columns is the pseudo type of `map<string, Columns>` in the `Statement` message.
+def generate_analysis_steps(datasource:str, columns:Columns):
+    """
+    `generate_analysis_steps` generates a `select` statement to get necessary aggregate data for feature columns
+    """
+    pass
+  
+# ColumnsDAG is a structure to be designed in the detailed design doc.
+def generate_feature_columns(columns: ColumnsDAG):
+    """
+    `generate_feature_columns` generates python feature columns objects
+    """
+    pass
+```
+
+`generate_analysis_steps` does at least the following tasks:
+
+- Call the standard module `ast` to generate ASTs for the function call expressions in `columns`
```

Comment:
Do you mean `ast` python module? I think `AS` keyword in COLUMN clause will break the execution of `ast` module.

Reply:
I propose to use the python `ast` module to generate ASTs because `COLUMN`s are actually calling python functions, we don't have to rewrite a parser to parse the *python style function call syntax*.

Similarly, I propose that we don't invent the new syntax `a_column(b_column() as c)` because it's neither SQL-stylish nor a valid function call expression. `b_column() as c, a(c)` is more explicit, more SQL-stylish and easier to parse.

Reply:
Do we use the standard python `ast` module here? Does the standard python `ast` module support to parse SQL statement, or user-defined DSL grammar? I remember the standard python `ast` module only supports to parse python code.

Reply:
Yes, this part proposes to use a python-like function call syntax for the `COLUMN` clause.